### ISSUE TITLE
[py] Crash when a window_switching_tests fails due to an unexpected exception

### ISSUE THREAD

**Issue by carlosgcampos** [SeleniumHQ/selenium#4636]
## Meta -
OS:  Linux

Selenium Version:  git master

This is happening since the switch to use a global driver instance. The problem is that when a test fails with an exception the pytest_exception_interact() is called, which quits the driver, and then the teardown part of the close_windows fixture tries to close the windows, but window_handles fails because the driver process is no longer running. It can easily be fixed just by handling any exception in close_windows after the yield, but I don't know if there's a batter way to do it, like checking if the driver is still running or something like that. If handling exceptions is the only way, that's how I fixed it locally, so I can just submit a PR.

## Expected Behavior -

The test that failed should be marked as F in the output, and the tests should continue starting a new driver after the failure without crashing.

## Actual Behavior -

Test is marked as F, and then it crashes

```test/selenium/webdriver/common/window_switching_tests.py ......F
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/main.py", line 110, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/main.py", line 146, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/main.py", line 169, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
INTERNALERROR>     return _wrapped_call(hook_impl.function(*args), self.execute)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 254, in _wrapped_call
INTERNALERROR>     return call_outcome.get_result()
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 280, in get_result
INTERNALERROR>     _reraise(*ex)  # noqa
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__
INTERNALERROR>     self.result = func()
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
INTERNALERROR>     return _wrapped_call(hook_impl.function(*args), self.execute)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 254, in _wrapped_call
INTERNALERROR>     return call_outcome.get_result()
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 280, in get_result
INTERNALERROR>     _reraise(*ex)  # noqa
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__
INTERNALERROR>     self.result = func()
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/runner.py", line 68, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/runner.py", line 82, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/runner.py", line 168, in call_and_report
INTERNALERROR>     hook.pytest_exception_interact(node=item, call=call, report=report)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/cgarcia/src/git/selenium/py/conftest.py", line 129, in pytest_exception_interact
INTERNALERROR>     driver_instance.quit()
INTERNALERROR>   File "/home/cgarcia/src/git/selenium/py/selenium/webdriver/webkitgtk/webdriver.py", line 70, in quit
INTERNALERROR>     RemoteWebDriver.quit(self)
INTERNALERROR>   File "/home/cgarcia/src/git/selenium/py/selenium/webdriver/remote/webdriver.py", line 603, in quit
INTERNALERROR>     self.execute(Command.QUIT)
INTERNALERROR>   File "/home/cgarcia/src/git/selenium/py/selenium/webdriver/remote/webdriver.py", line 308, in execute
INTERNALERROR>     self.error_handler.check_response(response)
INTERNALERROR>   File "/home/cgarcia/src/git/selenium/py/selenium/webdriver/remote/errorhandler.py", line 194, in check_response
INTERNALERROR>     raise exception_class(message, screen, stacktrace)
INTERNALERROR> NoSuchWindowException: Message:
Traceback (most recent call last):
  File "/usr/bin/pytest", line 11, in <module>
    load_entry_point('pytest==3.2.1', 'console_scripts', 'pytest')()
  File "/usr/lib/python2.7/dist-packages/_pytest/config.py", line 58, in main
    return config.hook.pytest_cmdline_main(config=config)
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
    res = hook_impl.function(*args)
  File "/usr/lib/python2.7/dist-packages/_pytest/main.py", line 139, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/usr/lib/python2.7/dist-packages/_pytest/main.py", line 133, in wrap_session
    exitstatus=session.exitstatus)
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
    return _wrapped_call(hook_impl.function(*args), self.execute)
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 250, in _wrapped_call
    wrap_controller.send(call_outcome)
  File "/usr/lib/python2.7/dist-packages/_pytest/terminal.py", line 406, in pytest_sessionfinish
    outcome.get_result()
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 280, in get_result
    _reraise(*ex)  # noqa
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__
    self.result = func()
  File "/usr/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
    res = hook_impl.function(*args)
  File "/usr/lib/python2.7/dist-packages/_pytest/runner.py", line 56, in pytest_sessionfinish
    session._setupstate.teardown_all()
  File "/usr/lib/python2.7/dist-packages/_pytest/runner.py", line 472, in teardown_all
    self._pop_and_teardown()
  File "/usr/lib/python2.7/dist-packages/_pytest/runner.py", line 445, in _pop_and_teardown
    self._teardown_with_finalization(colitem)
  File "/usr/lib/python2.7/dist-packages/_pytest/runner.py", line 463, in _teardown_with_finalization
    self._callfinalizers(colitem)
  File "/usr/lib/python2.7/dist-packages/_pytest/runner.py", line 460, in _callfinalizers
    py.builtin._reraise(*exc)
  File "/usr/lib/python2.7/dist-packages/_pytest/runner.py", line 453, in _callfinalizers
    fin()
  File "/usr/lib/python2.7/dist-packages/_pytest/fixtures.py", line 754, in finish
    py.builtin._reraise(*e)
  File "/usr/lib/python2.7/dist-packages/_pytest/fixtures.py", line 748, in finish
    func()
  File "/usr/lib/python2.7/dist-packages/_pytest/fixtures.py", line 703, in teardown
    next(it)
  File "/home/cgarcia/src/git/selenium/py/test/selenium/webdriver/common/window_switching_tests.py", line 31, in close_windows
    for handle in driver.window_handles:
  File "/home/cgarcia/src/git/selenium/py/selenium/webdriver/remote/webdriver.py", line 629, in window_handles
    return self.execute(Command.W3C_GET_WINDOW_HANDLES)['value']
  File "/home/cgarcia/src/git/selenium/py/selenium/webdriver/remote/webdriver.py", line 306, in execute
    response = self.command_executor.execute(driver_command, params)
  File "/home/cgarcia/src/git/selenium/py/selenium/webdriver/remote/remote_connection.py", line 464, in execute
    return self._request(command_info[0], url, body=data)
  File "/home/cgarcia/src/git/selenium/py/selenium/webdriver/remote/remote_connection.py", line 526, in _request
    resp = opener.open(request, timeout=self._timeout)
  File "/usr/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 1228, in http_open
    return self.do_open(httplib.HTTPConnection, req)
  File "/usr/lib/python2.7/urllib2.py", line 1198, in do_open
    raise URLError(err)
urllib2.URLError: <urlopen error [Errno 111] Connection refused>
```

## Steps to reproduce -

Just force any test in window_switching_tests to raise an unexpected exception and run the tests.

**Comment by barancev** [comment]
For issues please provide a [concise reproducible test case](http://sscce.org/) and describe what results you are seeing and what results you expect.

See [CONTRIBUTING.md](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md#issue-contributions)

**Comment by carlosgcampos** [comment]
I don't know how to provide a self-contained test, I can submit a patch to make any of the window_switching_tests raise an exception, for example. I already posted the results I'm getting (it has a full backtrace of the crash) and the expected results (it shouldn't crash). If it's going to be easier I can simply create a PR with a fix and we can discuss the right fix there.

**Comment by barancev** [comment]
If you think that making a patch is easier than making a test -- yes, let's go with PR.

**Comment by carlosgcampos** [comment]
I've just created #4654. Should we close this issue then?

**Comment by barancev** [comment]
Let it be open until PR has been reviewed at least.

**Comment by lmtierney** [comment]
fixed via 1d39d6b
